plugins/modules/iam_role.py

```diff
@@ -545,52 +556,53 @@ def remove_instance_profiles(role_params, role):
         profile_name = profile['InstanceProfileName']
         try:
             if not module.check_mode:
-                client.remove_role_from_instance_profile(aws_retry=True, InstanceProfileName=profile_name, **role_params)
+                client.remove_role_from_instance_profile(aws_retry=True, InstanceProfileName=profile_name, RoleName=role_name)
                 if profile_name == role_name:
                     if delete_profiles:
                         try:
                             client.delete_instance_profile(InstanceProfileName=profile_name, aws_retry=True)
-                        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+                        except is_boto3_error_code('NoSuchEntityException'):
+                            pass
+                        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:  # pylint: disable=duplicate-except
                             module.fail_json_aws(e, msg="Unable to remove instance profile {0}".format(profile_name))
         except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
             module.fail_json_aws(e, msg="Unable to remove role {0} from instance profile {1}".format(role_name, profile_name))
 
 
-def destroy_role():
+def destroy_role(module, client):
 
     role_name = module.params.get('name')
-    role = get_role(role_name)
-    role_params = dict()
-    role_params['RoleName'] = role_name
-    boundary_params = dict(role_params)
-    boundary_params['PermissionsBoundary'] = ''
+    role = get_role(module, client, role_name)
 
     if role is None:
         module.exit_json(changed=False)
 
-    # Before we try to delete the role we need to remove any
-    # - attached instance profiles
-    # - attached managed policies
-    remove_instance_profiles(role_params, role)
-    update_managed_policies(role_params, role, [], True)
-
-    try:
-        if not module.check_mode:
-            client.delete_role(aws_retry=True, **role_params)
-    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-        module.fail_json_aws(e, msg="Unable to delete role")
+    if not module.check_mode:
+        # Before we try to delete the role we need to remove any
+        # - attached instance profiles
+        # - attached managed policies
+        # - embedded inline policies
+        remove_instance_profiles(module, client, role_name)
+        update_managed_policies(module, client, role_name, [], True)
+        remove_inline_policies(module, client, role_name)
```

Comment:
Do we need something like ´wait_iam_exists()` to take care, that the IAM role is empty?

Reply:
...I'm not sure if AWS ack the requests and if IAM takes some seconds until the role is empty....

Reply:
@markuman nope! That's tested here and never failed when I was testing

```
    - name: "Remove IAM Role (rapid)"
      iam_role:
        state: absent
        name: "{{ test_role }}"
      register: iam_role

    - name: "Remove IAM Role (rapid)"
      iam_role:
        state: absent
        name: "{{ test_role }}"
      register: iam_role_again

    - assert:
        that:
          - iam_role is changed
          - iam_role_again is not changed
```